arch/x86/core/x86_mmu.c

```diff
@@ -529,39 +695,117 @@ static void *page_pool_get(void *context)
+ * existing mappings. Not needed for establishing new mappings
+ */
+#define OPTION_FLUSH		BIT(1)
+
+/* Indicates that each PTE's permission bits should be restored to their
+ * original state when the memory was mapped. All other bits in the PTE are
+ * preserved.
+ */
+#define OPTION_RESET		BIT(2)
+
+/* Indicates that allocations from the page pool are allowed to instantiate
+ * new paging structures. Only necessary when establishing new mappings
+ * and the entire address space isn't pre-allocated.
+ */
+#define OPTION_ALLOC		BIT(3)
+
 /**
- * Low-level mapping function
+ * Low level page table update function for a virtual page
  *
- * Walk the provided page tables until we get to the PTE for the provided
- * virtual address, and set that to whatever is in 'entry_val'.
+ * For the provided set of page tables, update the PTE associated with the
+ * virtual address to a new value, using the mask to control what bits
+ * need to be preserved.
  *
- * If memory must be drawn to instantiate page table memory, it will be
- * obtained from the provided get_page() function. The function must
- * return a page-aligned pointer to a page-sized block of zeroed memory.
- * All intermediate tables have hard-coded flags of INT_FLAGS.
+ * It is permitted to set up mappings without the Present bit set, in which
+ * case all other bits may be used for OS accounting.
  *
- * Presumes we want to map a minimally sized page of CONFIG_MMU_PAGE_SIZE.
- * No support for mapping big pages yet; unclear if we will ever need it given
- * Zephyr's typical use-cases.
+ * Must call this with x86_mmu_lock held.
  *
- * TODO: There may be opportunities to optimize page table walks such as this
- * by using recusrive page table mappings, see for example
- * https://os.phil-opp.com/paging-implementation/#recursive-page-tables
- * May also help if we need fast virtual-to-physical translation outside of
- * the permanent memory mapping area.
+ * Common mask values:
+ *  MASK_ALL  - Update all PTE bits. Exitsing state totally discarded.
+ *  MASK_PERM - Only update permission bits. All other bits and physical
+ *              mapping preserved.
  *
- * @param ptables Top-level page tables pointer
- * @param virt Virtual address to set mapping
- * @param entry_val Value to set PTE to
- * @param get_page Function to draw memory pages from
- * @param ctx Context pointer to pass to get_page()
- * @retval 0 success
- * @retval -ENOMEM get_page() failed
+ * @param ptables Page tables to modify
+ * @param virt Virtual page table entry to update
+ * @param entry_val Value to update in the PTE (ignored if OPTION_RESET)
+ * @param mask What bits to update in the PTE (ignored if OPTION_RESET)
+ * @param options Control options, described above
+ * @retval 0 Success
+ * @retval -ENOMEM allocation required and no free pages (only if OPTION_ALLOC)
  */
 static int page_map_set(pentry_t *ptables, void *virt, pentry_t entry_val,
-			page_get_func_t get_page, void *ctx)
```

Comment:
You can now remove the typedef on `page_get_func_t`

Reply:
should be gone now.